src/components/App.tsx

```diff
@@ -207,6 +223,81 @@ export class App extends React.Component<any, AppState> {
     event.preventDefault();
   };
 
+  private initializeSocketClient = () => {
+    if (this.socket) {
+      return;
+    }
+    const roomMatch = getCollaborationLinkData(window.location.href);
+    if (roomMatch) {
+      this.socket = socketIOClient(SOCKET_SERVER);
+      this.roomID = roomMatch[1];
+      this.roomKey = roomMatch[2];
+      this.socket.on("init-room", () => {
+        this.socket && this.socket.emit("join-room", this.roomID);
+      });
+      this.socket.on(
+        "new-user-receive-update",
+        async (encryptedData: ArrayBuffer) => {
+          if (this.roomKey) {
+            const scene = await decryptSocketUpdateData(
+              encryptedData,
+              this.roomKey,
+            );
+            elements = scene.elements;
+            this.setState({});
+          }
+          if (this.socket) {
+            this.socket.off("new-user-receive-update");
+            this.socket.off("new-user-first-in-room");
+          }
+          this.socketInitialized = true;
+        },
+      );
+      this.socket.on("new-user-first-in-room", () => {
+        if (this.socket) {
+          this.socket.off("new-user-receive-update");
+          this.socket.off("new-user-first-in-room");
+        }
+        this.socketInitialized = true;
+      });
+      this.socket.on("room-user-count", (collaboratorCount: number) => {
+        this.setState({ collaboratorCount });
+      });
+      this.socket.on("new-user", async (socketID: string) => {
+        this.socketInitialized &&
+          this.socket &&
+          this.roomKey &&
+          this.socket.emit(
+            "new-user-send-update",
+            socketID,
+            await encryptSocketUpdateData(elements, this.state, this.roomKey),
+          );
+      });
+      this.socket.on("receive-update", async (encryptedData: ArrayBuffer) => {
+        if (this.roomKey) {
+          const scene = await decryptSocketUpdateData(
+            encryptedData,
+            this.roomKey,
+          );
+          elements = scene.elements;
+          this.setState({});
+        }
+        this.socketInitialized = true;
+      });
+      this.socket.on(
+        "receive-mouse-location",
```

Comment:
Let's have a single "broadcast" type here and read the actual type from the decrypted data.